File: bindings/csharp/Bison/Key.cs
// MIT License © 2025 Binary Dice Games
/**
 * @file Key.cs
 * @brief Runtime name hashing for the Bison C# binding, with a bounded
 *        memoization cache mirroring `bison/dynamic.py`'s `key()`.
 */

using System.Collections.Concurrent;

namespace Bdg.Bison;

/// <summary>
/// Computes the 32-bit FNV-1a hash of a field/method/class name -- the same
/// value the C++ side gets at compile time from the <c>"name"_key</c>
/// user-defined literal (a <c>constexpr</c> evaluation, so the hash never
/// costs anything at run time there).
///
/// C# has no equivalent of a `constexpr` string literal hash: every
/// <see cref="Dynamic"/> field/method access must therefore hash its name at
/// run time via <c>bison_key()</c>. <see cref="Of"/> memoizes that call in a
/// bounded cache for the same reason <c>bison.dynamic.key()</c> wraps it in
/// `functools.lru_cache(maxsize=4096)`: real callers draw field/method names
/// from a small, static, schema-defined set that gets hashed over and over
/// (every `obj.Field` access, every `proxy.Method()` call), so caching turns
/// most lookups into a dictionary hit instead of a string encode + P/Invoke
/// call. The cache is bounded rather than unbounded so a caller hashing
/// high-cardinality or externally-derived strings (e.g. untrusted input)
/// can't grow it without bound.
/// </summary>
public static class Key
{
    private const int MaxSize = 4096;

    // ConcurrentDictionary gives thread-safe reads/writes without a global
    // lock on the hot path (mirrors CPython's GIL-protected dict backing
    // lru_cache, but explicit here since .NET has no GIL to lean on).
    private static readonly ConcurrentDictionary<string, uint> Cache = new();

    /// <summary>Maximum number of distinct names memoized before eviction
    /// kicks in. Exposed so tests can assert on the bound, mirroring
    /// `dynamic_mod.key.cache_info().maxsize` in the Python test suite.</summary>
    public static int MaxCacheSize => MaxSize;

    /// <summary>Number of distinct names currently memoized.</summary>
    public static int CacheCount => Cache.Count;

    /// <summary>
    /// Returns the FNV-1a hash of <paramref name="name"/> (identical to
    /// <c>"name"_key</c> in C++ and <c>bison.key(name)</c> in Python).
    /// </summary>
    public static uint Of(string name)
    {
        if (Cache.TryGetValue(name, out var cached))
        {
            return cached;
        }

        var hash = Native.bison_key(name);

        // Bounded, not evicting-LRU: once full, stop admitting new entries
        // rather than paying for eviction bookkeeping on every insert. A
        // cache miss just falls back to the (already cheap) native call.
        if (Cache.Count < MaxSize)
        {
            Cache.TryAdd(name, hash);
        }
        return hash;
    }

    /// <summary>Clears every memoized entry. Test-only escape hatch mirroring
    /// <c>dynamic_mod.key.cache_clear()</c>.</summary>
    public static void ClearCache() => Cache.Clear();
}

File: bindings/android/bison-lib/src/main/java/com/bdg/bison/Key.java
// MIT License © 2025 Binary Dice Games
// Permission is hereby granted, free of charge, to use, copy, modify, and
// distribute this file. See the LICENSE file or
// https://opensource.org/licenses/MIT for details.

package com.bdg.bison;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Hashes field/class/method names into the {@code bison_hash} (32-bit
 * FNV-1a) identifiers the wire format and the C ABI key on.
 *
 * <p>C++'s {@code "name"_key} is a {@code constexpr} evaluated at compile
 * time; Java has no equivalent, so every lookup calls across the ABI
 * (bison_key()) the same way the Python and C# bindings do. Results are
 * memoized in a small bounded cache -- field/method names are drawn from a
 * static, schema-defined set reused across many calls, so this turns most
 * lookups into a map hit instead of a JNI round trip, while staying bounded
 * so a caller hashing high-cardinality strings can't grow it forever. Mirrors
 * Python's {@code functools.lru_cache(maxsize=4096)} and C#'s
 * {@code Key.Of()} cache.
 */
public final class Key {
  static {
    NativeLibrary.ensureLoaded();
  }

  private static final int MAX_ENTRIES = 4096;

  // Access-order LinkedHashMap doubling as an LRU cache via removeEldestEntry.
  private static final Map<String, Integer> CACHE =
      new LinkedHashMap<String, Integer>(16, 0.75f, true) {
        @Override
        protected boolean removeEldestEntry(Map.Entry<String, Integer> eldest) {
          return size() > MAX_ENTRIES;
        }
      };

  /** Hashes {@code name}, matching the internal C++ {@code "name"_key} / {@code hash()}. */
  public static synchronized int of(String name) {
    if (name == null || name.isEmpty()) return 0;
    Integer cached = CACHE.get(name);
    if (cached != null) return cached;
    int hash = nativeKey(name);
    CACHE.put(name, hash);
    return hash;
  }

  private static native int nativeKey(String name);

  private Key() {}
}
